What's new for preview 8

ajcvickers opened the pull request.

ajcvickers commented:
@dotnet/efteam Should we include many-to-many here? I'm somewhat reluctant because there are a lot of additional tweaks/fixes that have gone in after we branched.

ajcvickers commented on entity-framework/core/what-is-new/ef-core-5.0/whatsnew.md:
```diff
@@ -14,6 +14,202 @@ This page does not duplicate the [plan for EF Core 5.0](xref:core/what-is-new/ef
+
+Documentation is tracked by issue [#1979](https://github.com/dotnet/EntityFramework.Docs/issues/1979).
+
+### Migrations: Rebuild SQLite tables
```
@bricelam I was able to get the example to work with the daily builds, but I can't make it work with preview 8. Do we want to bump rebuilds to the rc1 what's new?

bricelam replied:
Weird. What’s the error? I seem to remember the bugs in preview8 being pretty corner case. But yeah, waiting for rc1 is also fine with me.

ajcvickers replied:
Same as I got with 3.1:
```
System.NotSupportedException: SQLite does not support this migration operation ('DropColumnOperation'). For more information, see http://go.microsoft.com/fwlink/?LinkId=723262.
   at Microsoft.EntityFrameworkCore.Migrations.SqliteMigrationsSqlGenerator.Generate(DropColumnOperation operation, IModel model, MigrationCommandListBuilder builder, Boolean terminate)
   at Microsoft.EntityFrameworkCore.Migrations.MigrationsSqlGenerator.<>c.<.cctor>b__71_12(MigrationsSqlGenerator g, MigrationOperation o, IModel m, MigrationCommandListBuilder b)
   at Microsoft.EntityFrameworkCore.Migrations.MigrationsSqlGenerator.Generate(MigrationOperation operation, IModel model, MigrationCommandListBuilder builder)
   at Microsoft.EntityFrameworkCore.Migrations.MigrationsSqlGenerator.Generate(IReadOnlyList`1 operations, IModel model)
   at Microsoft.EntityFrameworkCore.Migrations.SqliteMigrationsSqlGenerator.Generate(IReadOnlyList`1 operations, IModel model)
   at Microsoft.EntityFrameworkCore.Migrations.Internal.Migrator.GenerateUpSql(Migration migration)
   at Microsoft.EntityFrameworkCore.Migrations.Internal.Migrator.<>c__DisplayClass15_2.<GetMigrationCommandLists>b__2()
   at Microsoft.EntityFrameworkCore.Migrations.Internal.Migrator.Migrate(String targetMigration)
   at Microsoft.EntityFrameworkCore.Design.Internal.MigrationsOperations.UpdateDatabase(String targetMigration, String contextType)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.UpdateDatabaseImpl(String targetMigration, String contextType)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.UpdateDatabase.<>c__DisplayClass0_0.<.ctor>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.Execute(Action action)
```

bricelam replied:
Hmm, doesn’t look like any of it is in preview8

bricelam replied:
That sounds right, I think I missed the cutoff date by a day or two.

ajcvickers replied:
Thanks for checking. I'll remove it here, and keep in the back pocket for rc1.

bricelam replied:
Well, the commit is in the branch...

bricelam replied:
I confirmed that this works for me on the 5.0.0-preview.8.20407.4 bits. Keep it in these release notes.

ajcvickers replied:
What feed did you get that build from? I was using an older build.

bricelam replied:
*.NET 5 preview8 build* email thread from Doug.

ajcvickers commented on entity-framework/core/what-is-new/ef-core-5.0/whatsnew.md:
```diff
@@ -14,6 +14,202 @@ This page does not duplicate the [plan for EF Core 5.0](xref:core/what-is-new/ef
+
+Documentation is tracked by issue [#2583](https://github.com/dotnet/EntityFramework.Docs/issues/2583).
+
+### Flexible type/function mapping
```
@AndriySvyryd @smitpatel Could one or both of you write either one combined or multiple different sections for the issues listed below? I'm not sure quite how to approach explaining this to people.

AndriySvyryd replied:
https://github.com/dotnet/efcore/issues/21415 Can be ignored, this is very low-level

AndriySvyryd commented:
> Should we include many-to-many here?

Anyone following the issue already knows that some of it it there. I think that's enough exposure for now

bricelam commented on entity-framework/core/what-is-new/ef-core-5.0/whatsnew.md:
```diff
@@ -14,6 +14,202 @@ This page does not duplicate the [plan for EF Core 5.0](xref:core/what-is-new/ef
+);
+
+CREATE TABLE [Cats] (
+    [Id] int NOT NULL,
+    [EdcuationLevel] nvarchar(max) NULL,
+    CONSTRAINT [PK_Cats] PRIMARY KEY ([Id])
```
If this is going in the blog post, we may want to preemptively include the foreign key constraints here, so we don't get an unnecessary tweetstorm.

ajcvickers replied:
I would rather not put in the blog post things that aren't in preview 8. But given I was using the wrong build, this is maybe not what is in preview 8?

If do still want to included this and this is what's generated for preview 8, then we could indicate that this is the SQL we generate in preview 8, but that it is changing in RC1.

bricelam replied:
Thoughts, @AndriySvyryd?

ajcvickers commented:
@dotnet/efteam:
* Updated TPT to include the constraints
* Added TVF mapping (/cc @pmiddleton - I started from what you wrote--it was very helpful thanks!)
* Added some words on query roots

ajcvickers commented:
@JeremyLikness I think this has everything in it now.

smitpatel commented on entity-framework/core/what-is-new/ef-core-5.0/whatsnew.md:
```diff
@@ -14,6 +14,350 @@ This page does not duplicate the [plan for EF Core 5.0](xref:core/what-is-new/ef
+
+This feature was contributed from the community by [@pmiddleton](https://github.com/pmiddleton). Many thanks for the contribution!
+
+EF Core 5.0 includes first-class support for mapping .NET methods table-valued functions (TVFs). These functions can then be used in LINQ queries where additional composition on the results of the function will also be translated to SQL.
```
```suggestion
EF Core 5.0 includes first-class support for mapping .NET methods to table-valued functions (TVFs). These functions can then be used in LINQ queries where additional composition on the results of the function will also be translated to SQL.
```

smitpatel commented on entity-framework/core/what-is-new/ef-core-5.0/whatsnew.md:
````diff
@@ -14,6 +14,350 @@ This page does not duplicate the [plan for EF Core 5.0](xref:core/what-is-new/ef
+modelBuilder.Entity(typeof(Report)).HasNoKey();
+```
+
+Notice that `Report` has no primary key and so much be configured as such.
````
```suggestion
Notice that `Report` has no primary key and it is configured as such.
```

smitpatel commented on entity-framework/core/what-is-new/ef-core-5.0/whatsnew.md:
````diff
@@ -14,6 +14,350 @@ This page does not duplicate the [plan for EF Core 5.0](xref:core/what-is-new/ef
+    => FromExpression(() => GetReports(managerId));
+```
+
+This method uses a parameter and return type that match the TVF defined above. The is then added to the EF Core model in OnModelCreating:
````
```suggestion
This method uses a parameter and return type that match the TVF defined above. The method is then added to the EF Core model in OnModelCreating:
```

pmiddleton commented on entity-framework/core/what-is-new/ef-core-5.0/whatsnew.md:
````diff
@@ -183,12 +190,153 @@ Notice that:
+    => FromExpression(() => GetReports(managerId));
+```
+
+This method uses a parameter and return type that match the TVF defined above. The is then added to the EF Core model in OnModelCreating:
````
It is then

ajcvickers commented:
@dotnet/efteam Can I get approval on this so we can push to live for the release today.

bricelam commented on entity-framework/core/what-is-new/ef-core-5.0/whatsnew.md:
````diff
@@ -14,6 +14,350 @@ This page does not duplicate the [plan for EF Core 5.0](xref:core/what-is-new/ef
+For example, consider this TVF defined in a SQL Server database:
+
+```sql
+create FUNCTION GetReports(@employeeId int)
+RETURNS @reports TABLE
+(
+	Name nvarchar(50) not null,
+	IsDeveloper bit not null
+)
+AS
+begin
+	WITH cteEmployees AS
+	(
+		SELECT id, name, managerId, isDeveloper
+		FROM employees
+		WHERE id = @employeeId
+		UNION ALL
+		SELECT e.id, e.name, e.managerId, e.isDeveloper
+		FROM employees e
+		INNER JOIN cteEmployees cteEmp ON cteEmp.id = e.ManagerId
+	)
+
+	insert into @reports
+	select name, isDeveloper
+	FROM cteEmployees
+	where id != @employeeId
+
+	return
+end
````
```suggestion
CREATE FUNCTION GetReports(@employeeId int)
RETURNS @reports TABLE
(
	Name nvarchar(50) NOT NULL,
	IsDeveloper bit NOT NULL
)
AS
BEGIN
	WITH cteEmployees AS
	(
		SELECT Id, Name, ManagerId, IsDeveloper
		FROM Employees
		WHERE Id = @employeeId
		UNION ALL
		SELECT e.Id, e.Name, e.ManagerId, e.IsDeveloper
		FROM Employees e
		INNER JOIN cteEmployees cteEmp ON cteEmp.Id = e.ManagerId
	)

	INSERT INTO @reports
	SELECT Name, IsDeveloper
	FROM cteEmployees
	WHERE Id != @employeeId

	RETURN
END
```

ajcvickers replied:
I left the casing as Paul had it, but I agree this aligns better with how we usually show SQL.

bricelam commented on entity-framework/core/what-is-new/ef-core-5.0/whatsnew.md:
````diff
@@ -14,6 +14,350 @@ This page does not duplicate the [plan for EF Core 5.0](xref:core/what-is-new/ef
+This method uses a parameter and return type that match the TVF defined above. The method is then added to the EF Core model in OnModelCreating:
+
+```c#
+modelBuilder.HasDbFunction(() => GetReports(0));
````
```suggestion
modelBuilder.HasDbFunction(() => GetReports(default));
```

ajcvickers replied:
I wondered about doing this. Probably still going to confuse some people. :-)
